### DISCUSSION TITLE
How should ExNVR use Cloud Storage CRUD?

### DISCUSSION THREAD

**Discussion by marcoherbst** [evercam/ex_nvr#370] [General]
My understanding is that ExNVR can have an attached S3 compatible folder. So far it seems unclear how this would be best used. I thought it might make sense to open a discussion.

- Timelapse extraction
- Clip extraction.
- Periodic snapshot upload.
- Upload clip on motion (or other event)
- ??

**Comment by Abdelhak-Bahri** [comment]
Exnvr may need S3 as a hard drive, so it will be able to store RAW video footage. When we need to download a video we will ask exnvr, and it fetches it from harddrive/s3 or whatever, because exnvr knows where it is and how to build it by merging the appropriate chunks.

I think a backend (media API) will always be responsible for managing other assets on S3-like buckets: which bucket/folder and how to name the file, which policy to apply, data retention policy, because in this case Media API knows better where to store them, who is the user, what project, who will be able to access it.

Periodic snapshots should be pushed to a gateway ( or let's say media API for now) that will manage how to store them.

Clips and time lapses will be generated based on a request from the dashboard/ Media API, so technically the media API will know better how to store them.
storing these files to S3 by exnvr can be an async alternative solution, in this case, media API will store metadata only and should know how these files are stored or should ask exnvr each time to find them.
I prefer that media API calls exnvr and gets the data and stores it and finds it when needed (also when edge device is decommissioned)

Uploading a clip on events is similar to periodic snapshots, it needs a gateway.


Just a preference, data related to customers or Evercam features should be managed by Media API.

**Comment by marcoherbst** [comment]
A good reference here is Kerberos Vault https://doc.kerberos.io/vault/first-things-first/#storage-providers 

Today I discussed with Billal the merits of offering s3 upload footage, as archives, when the edge storage is getting full. We agreed this was not likely to be helpful other than in an emergency due to SIM data limits.

At the same time we discussed whether uploaded video would be treated differently. As "archives" or somesuch. My opinion is that there is no reason to think of it differently to any other storage resource. We can treat s3 just like we would treat local storage. (?). In fact, for simplicity, we could choose that all storage on an ExNVR device is on an S3 compatible filesystem. (?)

**Comment by jakew009** [comment]
Fwiw's I'm really not sure that having ex_nvr talk directly to S3 is a good idea at all. How are you going to manage the S3 keys? Create a separate bucket for every customer with it's own keys? How do you stop an S3 key getting leaked and finding yourself with an enormous bill?
